merlin/models/torch/models/base.py

```diff
@@ -0,0 +1,226 @@
+    schema: Schema, optional
+        A Merlin schema. Default is None.
+    optimizer: torch.optim.Optimizer, optional
+        A PyTorch optimizer from the PyTorch library (or any custom optimizer
+        that follows the same API). Default is Adam optimizer.
+
+    Example usage
+    -------------
+    >>> model = Model(
+    ...    TabularInputBlock(schema),
+    ...    MLPBlock([32, 16]),
+    ...    BinaryOutput(schema.select_by_tag(Tags.TARGET).first),
+    ... )
+    ... trainer = Trainer(max_epochs=1)
+    ... with Loader(dataset, batch_size=16) as loader:
+    ...     model.initialize(loader)
+    ...     trainer.fit(model, loader)
+    """
+
+    def __init__(
+        self,
+        *blocks: nn.Module,
+        schema: Optional[Schema] = None,
+        optimizer=torch.optim.Adam,
+    ):
+        """Initializes `Model` class"""
+        super().__init__()
+        self.schema = schema
+
+        self.pre = BlockContainer(name="pre")
+        self.blocks = BlockContainer(name="blocks")
+        for block in blocks:
+            self.blocks.append(block)
+        self.post = BlockContainer(name="post")
+
+        self.optimizer = optimizer
+
+    def initialize(self, data: Union[Dataset, Loader, Batch]):
+        """Initializes the model based on a given data set."""
+        return initialize(self, data)
+
+    def forward(
+        self, inputs: Union[torch.Tensor, Dict[str, torch.Tensor]], batch: Optional[Batch] = None
+    ):
+        """Performs a forward pass through the model."""
+        outputs = inputs
+        for pre in self.pre.values:
+            outputs = pre(outputs, batch=batch)
+        for block in self.blocks.values:
+            outputs = block(outputs, batch=batch)
+        for post in self.post.values:
+            outputs = post(outputs, batch=batch)
+        return outputs
+
+    def training_step(self, batch, batch_idx):
+        """Performs a training step with a single batch."""
+        del batch_idx
+        inputs, targets = batch
+        predictions = self(inputs)
+
+        loss_and_metrics = compute_loss(predictions, targets, self.model_outputs())
+        for name, value in loss_and_metrics.items():
+            self.log(f"train_{name}", value)
+
+        return loss_and_metrics["loss"]
+
+    def configure_optimizers(self):
+        """Configures the optimizer for the model."""
+        return self.optimizer(self.parameters())
+
+    def model_outputs(self) -> List[ModelOutput]:
+        """Finds all instances of `ModelOutput` in the model."""
+        return module_utils.find_all_instances(self, ModelOutput)
+
+    def first(self) -> nn.Module:
+        """Returns the first block in the model."""
+        return self.blocks.values[0]
+
+    def last(self) -> nn.Module:
+        """Returns the last block in the model."""
+        return self.blocks.values[-1]
+
+    def input_schema(self) -> Schema:
+        """Returns the input schema of the model."""
+        if self.schema:
```

Comment:
Maybe add a `TODO` here that we need to infer this from the `TabularInputBlock` when we get there.